MeshImporter/XFormImporter refactor (USDU-3 part 3)

judubu opened the pull request:
## Purpose of this PR

**Ticket/Jira #:**
https://jira.unity3d.com/browse/USDU-3
https://jira.unity3d.com/browse/USDU-10

<!-- Description of feature/change. Links to screenshots, design docs, user docs, etc. Remember reviewers may be outside your team, and not know your feature/area that should be explained more. -->

USD geometric Primvars can be given different kind of [interpolation ](https://graphics.pixar.com/usd/docs/api/class_usd_geom_primvar.html#Usd_InterpolationVals) across the uv space of a surface primitive:
- **constant**: one value per-prim
- **uniform**: one value per-face
- **vertex/varying**: one value per-vertex
- **facevarying**: one value per-vertex per-face

Each primvar can also be individually indexed for data-compression purpose.

This is an issue for Unity which requires all attributes of a mesh to be either interpolated per vertex or per-vertex per-face. 
As a consequence it forces us to sanitize the data on the way in before being pushed to GameObjects.

The sanitization was previously done in the MeshImporter right before setting the Unity mesh buffers which was single threaded and made it impossible to properly cross check all attributes arrays for their interpolation value.

This PR introduces a way to perform sanitization of the Mesh and XForm data right after deserialization while using the job system to lessen the impact of interpolation conversion. For that a new ISanitizable interface and 2 new classes SanitizedMeshSample and XFormMeshSample have been added and the jobified ReadAllJob struct have been modified to provide sanitization service after deserialization. 

The sanitization code takes care of:
    - xform conversion
    - triangulation
    - de-indexation of attributes
    - attribute scope conversion

As a consequence the MeshImporter has been greatly simplified.

The SkeletonImporter has been changed and the JointWeights and JointIndices are now deserialized/loaded by the MeshImporter to avoid aan extra mesh deserialization/sanitization.


## Testing

**Functional Testing status:**

<!-- Explanation of what's tested, how tested and existing or new automation tests. Can include manual testing by self and/or QA. Specify test plans. Rarely acceptable to have no testing.  -->
The sanitization service has almost 100% coverage but not the MeshImporter itself.

@cguer You should check the import with all the regular suspect (kitchen, human female, [walking teapot](https://pixar-community-production.s3.us-west-1.amazonaws.com/Siggraph/2020/AR_Walking_Teapot_2020_Straight_Line.usdz) ) and do a before after. 

**Performance Testing status:**

<!-- Could this PR affect performance? If so, what has been done to measure time taken / memory used etc? Can include new or existing performance tests. Also see [Ensuring Performance by Default](https://confluence.unity3d.com/display/DEV/Ensuring+Performance+by+Default). -->
I quickly profile the changes and they are a little bit slower both for static and animated data. Although GC has been greatly reduced. I'll update with the actual numbers after my PTO.

## Overall Product Risks
<!-- See Testing Status, Complexity and Halo Effect for your PR](https://confluence.unity3d.com/display/DEV/Testing+Status%2C+Complexity+and+Halo+Effect+for+your+PR). -->

**Complexity:**
<!-- (Minimal / Low / Medium / High) -->
High

**Halo Effect:**
<!-- (Minimal / Low / Medium / High) -->
High

## Additional information

**Note to reviewers:**

<!-- Info per person for what to focus on, or historical info to understand who have previously reviewed and coverage. Help them get context. -->

Check the SanitizedMeshSample for the actual sanitization code and the MeshImporter to see how the sanitized data is pushed to Unity Meshes.
There are some minor changes to the deserialization process to flag colors as actual primvars which give us access to the indices and interpolation values.
Check the test usda file to get a feel for the kind of crappy edge case that can come up and if I forgot anything.

**Reminder:**
<!-- Things to remember to do before finalizing this PR. -->
- [ ] Add entry in CHANGELOG.md _(if applicable)_

judubu committed:
Ignore abstract prims when getting paths by type.

judubu committed:
Update bindings binaries

judubu committed:
Update win-10 yamato image

judubu committed:
Add type alias suport for reverse binding

judubu committed:
Fix test

judubu committed:
MeshImporter/XFormImporter refactor

Sanitizing the USD Meshes and XForms to Unity standards is now done by the sample itself right after deserialization. This has the
benefit to run in jobs and to simplify the MeshImporter code greatly.
Sanitization code takes care of:
    - xform conversion
    - triangulation
    - de-indexation of attributes
    - attribute scope conversion

The SkeletonImporter has been changed and the JointWeights and JointIndices are now deserialized/loaded by the MeshImporter to gain a whole extra mesh deserialization/sanitization.

clusty reviewed (approved):
Only comment was around need for both: MeshSample and SanitizedMeshSample & friends classes.

judubu committed:
Add wrapper method to access typeAliases

judubu committed:
Public API cleanup + cosmetic changes

judubu committed:
Fix broken Handedness test

MeshImporter was always changing handedness regardless of the importOptions.

Cameras were not sanitized so I added a new SanitizedCameraSample

judubu committed:
Ignore SanitizeSampleTest in standalone tests

judubu committed:
Remove public modifier from ISanitize.Sanitize

judubu committed:
Merge branch 'dev' into USDU-3/03_mesh_importer_refactor

jcowles commented on package/com.unity.formats.usd/Dependencies/USD.NET.Unity/Geometry/MeshSampleBase.cs:
```diff
@@ -37,21 +36,12 @@ public class MeshSampleBase : PointBasedSample
         /// </summary>
         /// <remarks>
-        /// UV object types should be Vector{2,3,4}[], List of Vector{2,3,4}, or null.
+        /// UV object types should be Vector{2,3}[], List of Vector{2,3}, or null.
```
Why just 2,3 components?

judubu replied:
Because I could not find any reference to TexCoord4f. What are UV coords of type float4 by the way?

jcowles replied:
Unity has float4 uvs, which are useful sometimes for storing tangent data, etc

jcowles commented on package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Primvar.cs:
```diff
@@ -142,6 +142,10 @@ public Type GetValueType()
 
+        public bool IsArray => typeof(T).IsArray;
+
+        public int Length => IsArray && value != null ? (value as Array).Length : 0;
```
Nit pick, but I would add parens around (IsArray && value != null) for readability.

judubu replied:
Fixed in dd63d81

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/CubeImporter.cs:
```diff
@@ -78,7 +78,7 @@ public static class CubeImporter
                     // Constant color can just be set on the material.
-                    mat = options.materialMap.InstantiateSolidColor(usdCube.colors[0].gamma);
+                    mat = options.materialMap.InstantiateSolidColor(usdCube.colors.value[0].gamma);
                     Debug.Log("constant colors assigned");
```
I think we should remove this debug.log

judubu replied:
Fixed in dd63d81

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -464,44 +483,39 @@ public static GeometrySubsets ReadGeomSubsets(Scene scene, string path)
                     {
-                        int[] faceIndices = kvp.Value;
-                        int[] triangleIndices = new int[faceIndices.Length * 3];
+                        var faceIndices = kvp.Value;
```
Changing this to `var` makes the code harder to read, because we also don't know what types `kvp` holds.

judubu replied:
Added in dd63d81

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:
```diff
@@ -0,0 +1,473 @@
+            values = newValues.ToArray();
+        }
+
+        void ConvertInterpolationToFaceVarying<T>(ref T[] values, int[] vertexIndices, bool changeHandedness = false, TfToken interpolation = null)
```
I think it's worth adding a doc string for this function, since converting TO FaceVarying for import sounds strange.

I would expect things to be converted to Vertex/Varying interpolation for import.

judubu replied:
Added in dd63d81

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:
```diff
@@ -0,0 +1,473 @@
+        }
+
+        /// <summary>
+        /// Returns the interpolation of an data array based on the number of elements
```
It's worth noting that this method, in general, will not always be correct.

It can be correct for constant and uniform interpolation, but varying/vertex/facevarying will sometimes be ambiguous.

judubu replied:
You're right, indexed arrays could easily mess it up and they are probably frequentely used.  
Indexed Vertex/Varying arrays will be shorter that points.length and indexed FaceVarying arrays will be shorter than faceVertexIndices.length or even equal to points.length in a worst case scenario!
All those situation will cause the MeshImporter to fail creating a valid mesh.

judubu replied:
@jcowles Now the issue to solve is that we don't deserialize the interpolation type for normals and tangents (we don't do the indices either). Although we do it for the colors and uvs by defining them as Primvars. Normals and tangents are not technically USD primvars but do you see an issue if I set them as Primvars?

judubu replied:
Rereading the doc I had some of that wrong:
- normals can be "builtin" or primvars
- only primvar normals can be indexed
- builtin normals are subject to interpolation but not indexing

Primvar normals should normally take precedence on builtin normals but we dont support them at the moment. It leaves us with non indexed builtin normals for which the GuessInterpolation method should work.

@jcowles  What is the ambiguous situation you had in mind?

jcowles replied:
Indexed primvars for sure, but there are degenerate non-indexed cases, where you have four faces and all faces share the same verts - uniform, vertex/varying, face varying are not distinguishable. Or more common cases, e.g. if you have a single face, then vertex & face-varying are not distinguishable.

I think what you have is fine, but it's worth calling this out.

Other primitives (such as curves) have even more aliasing.

judubu replied:
Got it thanks for the update. Adding that as comment in the code.

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:
```diff
@@ -0,0 +1,473 @@
+[assembly: InternalsVisibleToAttribute("Unity.Formats.USD.Tests")]
+namespace Unity.Formats.USD
+{
+    public interface ISanitizable
```
I love this idea - it awesomely simplifies import code and isolates the transformation into concise operators

jcowles reviewed (approved):
Wow, this change is incredible :)

It's clear that a massive amount of effort went into this, really nice work!

I have a few minor comments, but overall it totally makes sense and is a great simplification.

cguer reviewed (approved):
QA approved 👍!

Tested on:
- 2021.2.0b10 (Windows + HDRP)

Sorry for the delay 🙈 ...

judubu committed:
Address PR minor comments

judubu committed:
Add a comment for edge cases that will make GuessInterpolation return incorrect results
